src/app/visual/datepicker/datepicker-visual.component.ts

```diff
@@ -93,4 +106,92 @@ export class DatepickerVisualComponent implements OnInit {
   public themeSettingsChange(themeSettings: SkyThemeSettings): void {
     this.themeSvc.setTheme(themeSettings);
   }
+
+  public onToggleCustomDatesClick(): void {
+    this.showCustomDates = !this.showCustomDates;
+  }
+
+  public onCalendarDateRangeChange(range: SkyCalendarDateRangeChangeEvent): void {
+    this.updateCustomDates(range);
+  }
+
+  /**
+   * Simulate an async call to fetch data and add custom "key" dates to the current date range.
+   */
+  public updateCustomDates(event: SkyCalendarDateRangeChangeEvent): void {
+    if (this.showCustomDates && event) {
+      const customDates: SkyDatepickerCustomDate[] = [];
+      customDates.push({
+        date: event.startDate,
+        disabled: false,
+        keyDate: true,
+        keyDateText: ['First date']
+      });
+
+      customDates.push({
+        date: this.getNextDate(event.startDate, 8),
+        disabled: false,
+        keyDate: true,
+        keyDateText: ['Important']
+      });
+
+      customDates.push({
+        date: this.getNextDate(event.startDate, 9),
+        disabled: false,
+        keyDate: true,
+        keyDateText: ['Also Important']
+      });
+
+      customDates.push({
+        date: this.getNextDate(event.startDate, 10),
+        disabled: true,
+        keyDate: true,
+        keyDateText: ['Disabled']
+      });
+
+      customDates.push({
+        date: this.getNextDate(event.startDate, 11),
+        disabled: true,
+        keyDate: false,
+        keyDateText: []
+      });
+
+      customDates.push({
+        date: this.getNextDate(event.startDate, 12),
+        disabled: false,
+        keyDate: true,
+        keyDateText: []
+      });
+
+      customDates.push({
+        date: this.getNextDate(event.startDate, 13),
+        disabled: false,
+        keyDate: true,
+        keyDateText: ['Multiple', 'Messages']
+      });
+
+      customDates.push({
+        date: event.endDate,
+        disabled: false,
+        keyDate: true,
+        keyDateText: ['Last date']
+      });
+
+      // Bind the stream to the event argument.
+      if (!event.customDates) {
```

Comment:
I'd prefer we use a new Observable each time instead of reusing one, since that's what the consumer would do in the real world. Reusing a stream may also result in multiple values being pushed to the stream on one subscription, which you wouldn't want. This should get you what you want:

```
event.customDates = of(customDates).pipe(delay(2000));
```

Reply:
Good tip. Done.